### DISCUSSION TITLE
``win32api.SetConsoleCtrlHandler`` not working

### DISCUSSION THREAD

**Discussion by gswifort** [CEXT-Dan/PyRx#105] [General]
When I run such code in python:

```py
import os
import time
from pathlib import Path

import win32api
import win32con

BASE_DIR = Path(__file__).parent


def on_exit(signal):
    with open(BASE_DIR / "signal.txt", "w") as f:
        evt = {
            win32con.CTRL_C_EVENT: "CTRL_C_EVENT",
            win32con.CTRL_BREAK_EVENT: "CTRL_BREAK_EVENT",
            win32con.CTRL_CLOSE_EVENT: "CTRL_CLOSE_EVENT",
            win32con.CTRL_LOGOFF_EVENT: "CTRL_LOGOFF_EVENT",
            win32con.CTRL_SHUTDOWN_EVENT: "CTRL_SHUTDOWN_EVENT",
        }.get(signal, None)
        print(evt, file=f)
    if evt:
        print(evt)
        return 0 if evt == "CTRL_C_EVENT" else 1


if __name__ == "__main__":
    win32api.SetConsoleCtrlHandler(on_exit, True)
    print(f"PID = {os.getpid()}")
    e = 1
    while True:
        try:
            print(e)
        except Exception:
            pass
        finally:
            e += 1
            time.sleep(1)

```
everything works as expected

```raw
PID = 8120
1
2
3
# ...
40
CTRL_BREAK_EVENT
41
42
# ...
CTRL_C_EVENT
Traceback (most recent call last):
# ...
^C
```

but when I run similar code in CAD environment by connecting to the console of existing python process it seems that `SetConsoleCtrlHandler` does not register the handler - ZWCAD is closed immediately:

```py
import sys
import traceback
from pathlib import Path

import win32api
import win32con
import win32console

BASE_DIR = Path(__file__).parent
PID = 28596  # your python PID


def on_exit(signal):
    with open(BASE_DIR / "signal_acad.txt", "w") as f:
        evt = {
            win32con.CTRL_C_EVENT: "CTRL_C_EVENT",
            win32con.CTRL_BREAK_EVENT: "CTRL_BREAK_EVENT",
            win32con.CTRL_CLOSE_EVENT: "CTRL_CLOSE_EVENT",
            win32con.CTRL_LOGOFF_EVENT: "CTRL_LOGOFF_EVENT",
            win32con.CTRL_SHUTDOWN_EVENT: "CTRL_SHUTDOWN_EVENT",
        }.get(signal, None)
        print(evt, file=f)
    if evt:
        print(evt)
        return 0 if evt == "CTRL_C_EVENT" else 1


def main():
    win32console.FreeConsole()
    win32console.AttachConsole(PID)
    win32api.SetConsoleCtrlHandler(on_exit, True)
    f_stdout = open("CONOUT$", "w")
    sys.stdout = sys.stderr = f_stdout
    print("Attached ...")


def PyRxCmd_doit():
    try:
        main()
        print("DONE")
    except Exception:
        traceback.print_exc()
```

**Comment by CEXT-Dan** [comment]
Hi,
Try AllocConsole, I highly doubt that will work well though you will be competing for stdin, stdout

```Python
import os
import sys
import traceback
from pathlib import Path

import psutil
import win32api
import win32con
import win32console

BASE_DIR = Path(__file__).parent

def on_exit(signal):
    with open(BASE_DIR / "signal_acad.txt", "w") as f:
        evt = {
            win32con.CTRL_C_EVENT: "CTRL_C_EVENT",
            win32con.CTRL_BREAK_EVENT: "CTRL_BREAK_EVENT",
            win32con.CTRL_CLOSE_EVENT: "CTRL_CLOSE_EVENT",
            win32con.CTRL_LOGOFF_EVENT: "CTRL_LOGOFF_EVENT",
            win32con.CTRL_SHUTDOWN_EVENT: "CTRL_SHUTDOWN_EVENT",
        }.get(signal, None)
        print(evt, file=f)
    if evt:
        print(evt)
        return 0 if evt == "CTRL_C_EVENT" else 1


def main():
    win32console.FreeConsole()
    win32console.AllocConsole()
    win32api.SetConsoleCtrlHandler(on_exit, True)
    f_stdout = open("CONOUT$", "w")
    sys.stdout = sys.stderr = f_stdout
    print("Attached ...")


def PyRxCmd_doit():
    try:
        main()
        proccess = psutil.Process(os.getpid())
        print(proccess.name())
        print("DONE")
    except Exception:
        traceback.print_exc()
```

**Reply by gswifort** [reply]
Yes, I tried `AllocConsole` first - it doesn't work.

Ok, it's not that important.. I just wanted to avoid accidentally closing the console.

**Reply by CEXT-Dan** [reply]
Interested to unpack 3.13’s console as it’s supposed to have a bunch of new features. 
wx.py.shell works because it can capture stdin with key events. 

Another interesting thing to take note of is how debugpy works. it can somehow bypass the host application and communicate with the python interpreter through some special protocol.  

then there's VS Code's terminal, I’m sure there’s enough open-source stuff out there to build a cool cad python terminal, just a lot of work.

What do you use the console for? Is it cad related?

**Reply by gswifort** [reply]
I use the console to display test results (`pytest`), the results in the CAD console are unreadable.
Additionally, the CAD console does not display stdout so until it finishes the command, so if you are performing some time-consuming task, you do not see what is happening (no logs during work).

**Reply by gswifort** [reply]
Also for readability reasons I would prefer to run python console in terminal instead of wx palette. Then you can use ready solutions like `ptpython`, we just have to solve this problem: https://github.com/CEXT-Dan/PyRx/discussions/91#discussioncomment-10035328

**Comment by gswifort** [comment]
Can we get back to this topic? Do you see a way to catch ``CTRL_BREAK_EVENT/CTRL_C_EVENT``, currently when you click Ctrl+C in the console window the host closes immediately, the same when you close the console window.

**Reply by CEXT-Dan** [reply]
The only way I can think of is to try win32gui.CallWindowProc or some sort of PreTranslateMessage. This should not be hooked into the host, Ctrl+C is copy
